models/seir/model_mod.f90

```diff
@@ -0,0 +1,455 @@
+x3 = delta_t * dx
+
+! 4th step
+xi = x + delta_t * dx
+call seir_eqns(xi, dx)  
+x4 = delta_t * dx
+
+! Compute new value for x
+x = x + x1/6.0_r8 + x2/3.0_r8 + x3/3.0_r8 + x4/6.0_r8
+
+end subroutine adv_1step
+
+!------------------------------------------------------------------
+! SEIR Model Equations
+! The following extended SEIR Model with Vaccination 
+! is adaopted from Ghostine et al. (2021):
+! Ghostine, R., Gharamti, M., Hassrouny, S and Hoteit, I 
+! "An Extended SEIR Model with Vaccination for Forecasting
+! the COVID-19 Pandemic in Saudi Arabia Using an Ensemble
+! Kalman Filter" Mathematics 2021, 9, 636.
+! https://dx.doi.org/10.3390/math9060636 
+
+subroutine seir_eqns(x, fx)
+
+! State: x = [S, E, I, Q, R, D, V]
+
+real(r8), intent(in)  :: x(:)
+real(r8), intent(out) :: fx(:)
+
+fx(1) = theta - beta * x(1) * x(3) - alpha * x(1) - mu * x(1)
+fx(2) = beta * x(1) * x(3) - gama * x(2) + sigma * beta * x(7) * x(3) - mu * x(2)
+fx(3) = gama * x(2) - delta * x(3) - mu * x(3)
+fx(4) = delta * x(3) - (1.0_r8 - kappa) * lambda * x(4) - kappa * rho * x(4) - mu * x(4)
+fx(5) = (1.0_r8 - kappa) * lambda * x(4) - mu * x(5)
+fx(6) = kappa * rho * x(4)
+fx(7) = alpha * x(1) - sigma * beta * x(7) * x(3) - mu * x(7)
+
+end subroutine seir_eqns
+
+!------------------------------------------------------------------
+! Perturbs a model state for generating initial ensembles.
+! Returning interf_provided .true. means this code has
+! added uniform small independent perturbations to a
+! single ensemble member to generate the full ensemble.
+subroutine pert_model_copies(state_ens_handle, ens_size, pert_amp, interf_provided)
+
+type(ensemble_type), intent(inout) :: state_ens_handle
+integer,   intent(in) :: ens_size
+real(r8),  intent(in) :: pert_amp
+logical,  intent(out) :: interf_provided
+
+integer                  :: i,j, num_my_grid_points
+integer(i8), allocatable :: my_grid_points(:)
+type(location_type)      :: location
+integer                  :: var_type
+real(r8)                 :: rng
+
+interf_provided = .true.
+
+call init_random_seq(random_seq, my_task_id()+1)
+! if we are running with more than 1 task, then
+! we have all the ensemble members for a subset of
+! the model state.  which variables we have are determined
+! by looking at the global index number into the state vector.
+
+! how many grid points does my task have to work on?
+! and what are their indices into the full state vector?
+num_my_grid_points = get_my_num_vars(state_ens_handle)
+allocate(my_grid_points(num_my_grid_points))
+call get_my_vars(state_ens_handle, my_grid_points)
+
+do i=1,num_my_grid_points
+    call get_state_meta_data(my_grid_points(i), location, var_type)
+    
+    ! Lognormal Distribution
+    do j= 1, ens_size
+         rng = pert_size * random_gaussian(random_seq, 0.0_r8, 1.0_r8)
+         state_ens_handle%copies(j, i) = &
+            state_ens_handle%copies(j, i) * exp(rng)
+    end do
+end do
+
+deallocate(my_grid_points)
```

Comment:
It looks like the perturb does not depend on the state_index var_type or location, so you can remove the get_state_meta data,  get_my_vars and my_grid_points.

```suggestion
num_my_grid_points = get_my_num_vars(state_ens_handle)

do i=1,num_my_grid_points
    
    ! Lognormal Distribution
    do j= 1, ens_size
         rng = pert_size * random_gaussian(random_seq, 0.0_r8, 1.0_r8)
         state_ens_handle%copies(j, i) = &
            state_ens_handle%copies(j, i) * exp(rng)
    end do
end do

```

Reply:
Agreed, thanks for catching this Helen.